### PR TITLE
meson: 0.64.1 -> 1.0.0

### LINKED ISSUE #211842: meson-1.0.0 is out (mesa 23.0.0-rcX needs it)
Hi,

I can't see an update on master or a pull request.

Mesa 23.0.0 seems to need a meson feature that appears to only be in meson-1.0.0 which is out. I realize bumping meson is probably a large task though, so I'm raising a flag for it now.

cc: @raboof @jtojnar (recent committers to our meson package)

### PR DESCRIPTION
**Pull request by NickCao**
###### Description of changes
Changes: https://github.com/mesonbuild/meson/compare/0.64.1...1.0.0
Fixes https://github.com/NixOS/nixpkgs/issues/211842

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NickCao** [comment]
Built all the way to `systemd` without issue.

**Comment by NickCao** [comment]
Same for `nixosTests.simple`.

**Comment by AndersonTorres** [comment]
What about Darwin Intel?

**Comment by NickCao** [comment]
It's failing due to timeout.

**Comment by AndersonTorres** [comment]
Looks like me at #206799

**Comment by NickCao** [comment]
Despite the major version bump, the changes does not look that scary, I think we are good to give it a try.

**Comment by NickCao** [comment]
The situation is far better than I originally anticipated.

File: pkgs/development/tools/build-managers/meson/default.nix [hunk 1]
```diff
@@ -65,13 +67,23 @@ python3.pkgs.buildPythonApplication rec {
...
+    substituteInPlace "test cases/native/8 external program shebang parsing/script.int.in" \
+      --replace /usr/bin/env ${coreutils}/bin/env
+    # requires git, creating cyclic dependency
+    rm -r "test cases/common/66 vcstag"
+    # requires glib, creating cyclic dependency
+    rm -r "test cases/linuxlike/6 subdir include order"
+    rm -r "test cases/linuxlike/9 compiler checks with dependencies"
+    # requires cmake, creating cyclic dependency
+    substituteInPlace "test cases/failing/111 empty fallback/test.json" \
+      --replace "tried pkgconfig and cmake" "tried pkgconfig"
```
Comment (eli-schwartz) [inline pkgs/development/tools/build-managers/meson/default.nix:83]: Fixed upstream, thanks for the bug report.

**Review by eli-schwartz — COMMENTED** [review]
> Despite the major version bump, the changes does not look that scary, I think we are good to give it a try. But as there's still time ahead of the next staging-next cycle, I'd like to see if we can get the tests working.

:) It is a bump from a major version of 0 to a major version of 1, which is the sole time that semver doesn't say anything about backwards compatibility breaks.

**Comment by AndersonTorres** [comment]
Please update

File: pkgs/development/tools/build-managers/meson/default.nix [hunk 2]
```diff
@@ -61,17 +63,31 @@ python3.pkgs.buildPythonApplication rec {
...
 
-  # Meson included tests since 0.45, however they fail in Nixpkgs because they
-  # require a typical building environment (including C compiler and stuff).
-  # Just for the sake of documentation, the next lines are maintained here.
-  doCheck = false;
   nativeCheckInputs = [ ninja pkg-config ];
+  checkInputs = [ zlib ];
   checkPhase = ''
-    python ./run_project_tests.py
+    patchShebangs "test cases"
```
Comment (AndersonTorres) [inline pkgs/development/tools/build-managers/meson/default.nix:80]: Mucho whitespace

**Review by trofi — APPROVED** [review]

**Comment by stephank** [comment]
I had to locally revert 022c01aa39f1df76eab961967ce48bdcbfe16684 while testing on `aarch64-darwin`, but am not on a clean branch. Just a heads up that this quite possibly broke darwin.

**Comment by NickCao** [comment]
We could just save it for later. The test is just a nice to have.

**Pull request merged**